### ISSUE TITLE
Grace and OpenTelemetry 1.9.0 = Exception.

### ISSUE THREAD

**Issue by secovel** [ipjohnson/Grace#317]
VERY minimal example attached.

Add Opentelemetry Tracing with OtlpExporter options.  Runs fine.  Add UserGrace(), get an exception.  Seems like Grace much be touching some objects it shouldn't during bootstrapping.

`fail: Microsoft.Extensions.Hosting.Internal.Host[11]
      Hosting failed to start
      System.NotSupportedException: Signal-specific AddOtlpExporter methods and the cross-cutting UseOtlpExporter method being invoked on the same IServiceCollection is not supported.
         at OpenTelemetry.Exporter.OpenTelemetryBuilderServiceProviderExtensions.EnsureNoUseOtlpExporterRegistrations(IServiceProvider serviceProvider)
         at OpenTelemetry.Trace.OtlpTraceExporterHelperExtensions.BuildOtlpExporterProcessor(IServiceProvider serviceProvider, OtlpExporterOptions exporterOptions, SdkLimitOptions sdkLimitOptions, ExperimentalOptions experimentalOptions, ExportProcessorType exportProcessorType, BatchExportProcessorOptions`1 batchExportProcessorOptions, Boolean skipUseOtlpExporterRegistrationCheck, Func`2 configureExporterInstance)
         at OpenTelemetry.Trace.OtlpTraceExporterHelperExtensions.BuildOtlpExporterProcessor(IServiceProvider serviceProvider, OtlpExporterOptions exporterOptions, SdkLimitOptions sdkLimitOptions, ExperimentalOptions experimentalOptions, Func`2 configureExporterInstance)
         at OpenTelemetry.Trace.OtlpTraceExporterHelperExtensions.<>c__DisplayClass2_0.<AddOtlpExporter>b__1(IServiceProvider sp)
         at OpenTelemetry.Trace.TracerProviderBuilderExtensions.<>c__DisplayClass8_0.<AddProcessor>b__0(IServiceProvider sp, TracerProviderBuilder builder)
         at OpenTelemetry.Trace.OpenTelemetryDependencyInjectionTracingServiceCollectionExtensions.ConfigureTracerProviderBuilderCallbackWrapper.ConfigureBuilder(IServiceProvider serviceProvider, TracerProviderBuilder tracerProviderBuilder)
         at OpenTelemetry.Trace.TracerProviderSdk..ctor(IServiceProvider serviceProvider, Boolean ownsServiceProvider)
         at lambda_method101(Closure, IExportLocatorScope, IDisposalScope, IInjectionContext)
         at Grace.DependencyInjection.Lifestyle.SingletonLifestyle.ProvideLifestyleExpression(IInjectionScope scope, IActivationExpressionRequest request, Func`2 activationExpression)
         at Grace.DependencyInjection.Impl.InstanceStrategies.DelegateBaseExportStrategy`1.CreateExpression(IInjectionScope scope, IActivationExpressionRequest request, ICompiledLifestyle lifestyle)
         at Grace.DependencyInjection.Impl.InstanceStrategies.BaseInstanceExportStrategy.GetActivationExpression(IInjectionScope scope, IActivationExpressionRequest request)
         at Grace.DependencyInjection.Impl.InstanceStrategies.BaseInstanceExportStrategy.GetActivationStrategyDelegate(IInjectionScope scope, IActivationStrategyCompiler compiler, Type activationType)
         at Grace.DependencyInjection.Impl.ActivationStrategyCompiler.LocateStrategyFromCollectionContainers(IInjectionScope scope, Type locateType, ActivationStrategyFilter consider, Object key, IInjectionContext injectionContext)
         at Grace.DependencyInjection.Impl.ActivationStrategyCompiler.FindDelegate(IInjectionScope scope, Type locateType, ActivationStrategyFilter consider, Object key, IInjectionContext injectionContext, Boolean checkMissing)
         at Grace.DependencyInjection.Impl.InjectionScope.InternalLocate(IExportLocatorScope scope, IDisposalScope disposalScope, Type type, ActivationStrategyFilter consider, Object key, IInjectionContext injectionContext, Boolean allowNull, Boolean isDynamic)
         at Grace.DependencyInjection.Impl.InjectionScope.Grace.DependencyInjection.IInjectionScope.LocateFromChildScope(IExportLocatorScope childScope, IDisposalScope disposalScope, Type type, Object extraData, ActivationStrategyFilter consider, Object key, Boolean allowNull, Boolean isDynamic)
         at Grace.DependencyInjection.Impl.ActivationStrategyDelegateCache.FallbackExecution(ImmutableHashTree`2 currentNode, Type type, IExportLocatorScope scope, Boolean allowNull, IInjectionContext context)
         at Grace.DependencyInjection.Impl.ActivationStrategyDelegateCache.ExecuteActivationStrategyDelegateAllowNull(Type type, IExportLocatorScope scope)
         at Grace.DependencyInjection.Impl.BaseExportLocatorScope.LocateOrDefault(Type type, Object defaultValue)
         at Grace.DependencyInjection.Extensions.GraceRegistration.GraceServiceProvider.GetService(Type serviceType)
         at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetService[T](IServiceProvider provider)
         at OpenTelemetry.Extensions.Hosting.Implementation.TelemetryHostedService.Initialize(IServiceProvider serviceProvider)
         at OpenTelemetry.Extensions.Hosting.Implementation.TelemetryHostedService.StartAsync(CancellationToken cancellationToken)
         at Microsoft.Extensions.Hosting.Internal.Host.<StartAsync>b__15_1(IHostedService service, CancellationToken token)
         at Microsoft.Extensions.Hosting.Internal.Host.ForeachService[T](IEnumerable`1 services, CancellationToken token, Boolean concurrent, Boolean abortOnFirstException, List`1 exceptions, Func`3 operation)
Unhandled exception. System.NotSupportedException: Signal-specific AddOtlpExporter methods and the cross-cutting UseOtlpExporter method being invoked on the same IServiceCollection is not supported.
   at OpenTelemetry.Exporter.OpenTelemetryBuilderServiceProviderExtensions.EnsureNoUseOtlpExporterRegistrations(IServiceProvider serviceProvider)
   at OpenTelemetry.Trace.OtlpTraceExporterHelperExtensions.BuildOtlpExporterProcessor(IServiceProvider serviceProvider, OtlpExporterOptions exporterOptions, SdkLimitOptions sdkLimitOptions, ExperimentalOptions experimentalOptions, ExportProcessorType exportProcessorType, BatchExportProcessorOptions`1 batchExportProcessorOptions, Boolean skipUseOtlpExporterRegistrationCheck, Func`2 configureExporterInstance)
   at OpenTelemetry.Trace.OtlpTraceExporterHelperExtensions.BuildOtlpExporterProcessor(IServiceProvider serviceProvider, OtlpExporterOptions exporterOptions, SdkLimitOptions sdkLimitOptions, ExperimentalOptions experimentalOptions, Func`2 configureExporterInstance)
   at OpenTelemetry.Trace.OtlpTraceExporterHelperExtensions.<>c__DisplayClass2_0.<AddOtlpExporter>b__1(IServiceProvider sp)
   at OpenTelemetry.Trace.TracerProviderBuilderExtensions.<>c__DisplayClass8_0.<AddProcessor>b__0(IServiceProvider sp, TracerProviderBuilder builder)
   at OpenTelemetry.Trace.OpenTelemetryDependencyInjectionTracingServiceCollectionExtensions.ConfigureTracerProviderBuilderCallbackWrapper.ConfigureBuilder(IServiceProvider serviceProvider, TracerProviderBuilder tracerProviderBuilder)
   at OpenTelemetry.Trace.TracerProviderSdk..ctor(IServiceProvider serviceProvider, Boolean ownsServiceProvider)
   at lambda_method101(Closure, IExportLocatorScope, IDisposalScope, IInjectionContext)
   at Grace.DependencyInjection.Lifestyle.SingletonLifestyle.ProvideLifestyleExpression(IInjectionScope scope, IActivationExpressionRequest request, Func`2 activationExpression)
   at Grace.DependencyInjection.Impl.InstanceStrategies.DelegateBaseExportStrategy`1.CreateExpression(IInjectionScope scope, IActivationExpressionRequest request, ICompiledLifestyle lifestyle)
   at Grace.DependencyInjection.Impl.InstanceStrategies.BaseInstanceExportStrategy.GetActivationExpression(IInjectionScope scope, IActivationExpressionRequest request)
   at Grace.DependencyInjection.Impl.InstanceStrategies.BaseInstanceExportStrategy.GetActivationStrategyDelegate(IInjectionScope scope, IActivationStrategyCompiler compiler, Type activationType)
   at Grace.DependencyInjection.Impl.ActivationStrategyCompiler.LocateStrategyFromCollectionContainers(IInjectionScope scope, Type locateType, ActivationStrategyFilter consider, Object key, IInjectionContext injectionContext)
   at Grace.DependencyInjection.Impl.ActivationStrategyCompiler.FindDelegate(IInjectionScope scope, Type locateType, ActivationStrategyFilter consider, Object key, IInjectionContext injectionContext, Boolean checkMissing)
   at Grace.DependencyInjection.Impl.InjectionScope.InternalLocate(IExportLocatorScope scope, IDisposalScope disposalScope, Type type, ActivationStrategyFilter consider, Object key, IInjectionContext injectionContext, Boolean allowNull, Boolean isDynamic)
   at Grace.DependencyInjection.Impl.InjectionScope.Grace.DependencyInjection.IInjectionScope.LocateFromChildScope(IExportLocatorScope childScope, IDisposalScope disposalScope, Type type, Object extraData, ActivationStrategyFilter consider, Object key, Boolean allowNull, Boolean isDynamic)
   at Grace.DependencyInjection.Impl.ActivationStrategyDelegateCache.FallbackExecution(ImmutableHashTree`2 currentNode, Type type, IExportLocatorScope scope, Boolean allowNull, IInjectionContext context)
   at Grace.DependencyInjection.Impl.ActivationStrategyDelegateCache.ExecuteActivationStrategyDelegateAllowNull(Type type, IExportLocatorScope scope)
   at Grace.DependencyInjection.Impl.BaseExportLocatorScope.LocateOrDefault(Type type, Object defaultValue)
   at Grace.DependencyInjection.Extensions.GraceRegistration.GraceServiceProvider.GetService(Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetService[T](IServiceProvider provider)
   at OpenTelemetry.Extensions.Hosting.Implementation.TelemetryHostedService.Initialize(IServiceProvider serviceProvider)
   at OpenTelemetry.Extensions.Hosting.Implementation.TelemetryHostedService.StartAsync(CancellationToken cancellationToken)
   at Microsoft.Extensions.Hosting.Internal.Host.<StartAsync>b__15_1(IHostedService service, CancellationToken token)
   at Microsoft.Extensions.Hosting.Internal.Host.ForeachService[T](IEnumerable`1 services, CancellationToken token, Boolean concurrent, Boolean abortOnFirstException, List`1 exceptions, Func`3 operation)
   at Microsoft.Extensions.Hosting.Internal.Host.StartAsync(CancellationToken cancellationToken)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.RunAsync(IHost host, CancellationToken token)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.RunAsync(IHost host, CancellationToken token)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.Run(IHost host)
   at Program.<Main>$(String[] args) in /Users/scovel/RiderProjects/WebApplication8/WebApplication8/Program.cs:line 63
`

**Comment by secovel** [comment]
Can't seem to attach a file, so here's Program.cs and the csproj file

`using System.Diagnostics;
using Grace.AspNetCore.Hosting;
using OpenTelemetry.Exporter;
using OpenTelemetry.Resources;
using OpenTelemetry.Trace;

var builder = WebApplication.CreateBuilder(args);

// Add services to the container.
builder.Services.AddEndpointsApiExplorer();
builder.Services.AddSwaggerGen();

//Comment out and the service starts
builder.Host.UseGrace();

#region Tracing

var resourceBuilder = ResourceBuilder.CreateDefault()
    .AddService(serviceName: "OpenTelementryTest.Service", serviceVersion: "1.0.0");

ActivitySource myActivitySource = new ActivitySource("OpenTelementryTest.Service", "1.0.0");
        
builder.Services.AddSingleton(myActivitySource);
        
builder.Services.AddOpenTelemetry().WithTracing(tracerProviderBuilder =>
{
    tracerProviderBuilder.SetSampler(new AlwaysOnSampler());
    tracerProviderBuilder.AddSource("OpenTelementryTest.Service")
        .AddSource(myActivitySource.Name)
        .SetResourceBuilder(resourceBuilder);

    tracerProviderBuilder.AddOtlpExporter(options =>
    {
        options.Endpoint = new Uri("https://live.dynatrace.com/api/v2/otlp/v1/traces");
        options.Headers = "Authorization=Api-Token XXXXXXXXXXX";
        options.Protocol = OtlpExportProtocol.HttpProtobuf;
    });
});

#endregion

var app = builder.Build();

var summaries = new[]
{
    "Freezing", "Bracing", "Chilly", "Cool", "Mild", "Warm", "Balmy", "Hot", "Sweltering", "Scorching"
};

app.MapGet("/weatherforecast", () =>
    {
        var forecast = Enumerable.Range(1, 5).Select(index =>
                new WeatherForecast
                (
                    DateOnly.FromDateTime(DateTime.Now.AddDays(index)),
                    Random.Shared.Next(-20, 55),
                    summaries[Random.Shared.Next(summaries.Length)]
                ))
            .ToArray();
        return forecast;
    })
    .WithName("GetWeatherForecast");

app.Run();

record WeatherForecast(DateOnly Date, int TemperatureC, string? Summary)
{
    public int TemperatureF => 32 + (int)(TemperatureC / 0.5556);
}`


`<Project Sdk="Microsoft.NET.Sdk.Web">

    <PropertyGroup>
        <TargetFramework>net8.0</TargetFramework>
        <Nullable>enable</Nullable>
        <ImplicitUsings>enable</ImplicitUsings>
        <InvariantGlobalization>true</InvariantGlobalization>
    </PropertyGroup>

    <ItemGroup>
        <PackageReference Include="Grace.AspNetCore.Hosting" Version="7.1.0" />
        <PackageReference Include="Microsoft.AspNetCore.OpenApi" Version="8.0.6" />
        <PackageReference Include="OpenTelemetry.Exporter.OpenTelemetryProtocol" Version="1.9.0" />
        <PackageReference Include="OpenTelemetry.Extensions.Hosting" Version="1.9.0" />
        <PackageReference Include="Swashbuckle.AspNetCore" Version="6.6.2" />
    </ItemGroup>

</Project>`

**Comment by secovel** [comment]
to narrow it down further.  It all worked with OpenTelemetry 1.7.0.  I think after that they turned on some "error checking" code that seems to be triggering the exception when Grace pokes some object.

Seems to be this code block that triggers the bad behavior:

tracerProviderBuilder.AddOtlpExporter(options =>
{
    options.Endpoint = new Uri("https://live.dynatrace.com/api/v2/otlp/v1/traces");
    options.Headers = "Authorization=Api-Token XXXXXXXXXXX";
    options.Protocol = OtlpExportProtocol.HttpProtobuf;
});

I need it to send Tracing to a specific Otel ingestion point for Dynatrace.  In the full code I have blocks for metrics and logging as well, each has it's own endpoint.

**Comment by secovel** [comment]
1.7.0 works, anything higher throws an exception.
1.7.0 now has a medium vulnerability so we're needing to upgrade soon.
Any help would be appreciated.

        <PackageReference Include="Grace.AspNetCore.Hosting" Version="7.1.0" />
        <PackageReference Include="OpenTelemetry.Exporter.OpenTelemetryProtocol" Version="1.7.0" />
        <PackageReference Include="OpenTelemetry.Extensions.Hosting" Version="1.7.0" />


Program.cs


```
using Grace.AspNetCore.Hosting;
using OpenTelemetry.Trace;

var builder = WebApplication.CreateBuilder(new WebApplicationOptions());

//Comment out and the service starts
builder.Host.UseGrace();

builder.Services.AddOpenTelemetry().WithTracing(tracerProviderBuilder =>
{
    tracerProviderBuilder.AddOtlpExporter();
});

var app = builder.Build();

app.MapGet("/hello", () =>
    {
        return "hello";
    })
    .WithName("hello");

app.Run();

```

**Comment by secovel** [comment]
Might be related to this?

https://github.com/open-telemetry/opentelemetry-dotnet/issues/5538

**Comment by secovel** [comment]
I opened an issue in the OTEL Git

Would you mind taking a look at this any maybe commenting?

https://github.com/open-telemetry/opentelemetry-dotnet/issues/5803

@ipjohnson

**Comment by secovel** [comment]
Talking to myself again...

The issue seems to be the auto-registration of `UseOtlpExporterRegistration`.  OpenTelemetry, starting with 1.8.x decided that having that registered meant you were USING it, and throws an error.  This is related to Grace's auto-registration.  They are going to fix OpenTelemetry to not make that assumption (since it breaks multiple other DI containers).  Meanwhile, waiting for a release I came up with the following solution:

```
builder.Host.UseGrace()
    .ConfigureContainer<IInjectionScope>(scope =>
    {
        scope.Configure(c =>
            c.ExcludeTypeFromAutoRegistration("OpenTelemetry.Exporter.UseOtlpExporterRegistration"));
    });
```

Basically tell Grace not to register that 'sensitive' class.

**Comment by jods4** [comment]
Yeah, auto-registration is a footgun that can create such problems. OpenTelemetry is not the only lib that does some funny registration business and can break when using Grace with default settings. See MassTransit #309 or Orleans #315.

Personally I'd advise to always disable it with `AutoRegisterUnknown = false`.

For better or worst, this is a long-time _feature_ of Grace, so I guess this issue can be closed. 
Maybe the only argument remaining would be: should the default value change to `false` for next major release (breaking change)?

**Comment by secovel** [comment]
Well, for my project we've been using autoregistration = true for so long, plus some Interception stuff, I know it would be breaking if the default were changed.  We moved from MS Unity which acts the same.  

The first thing I tried today to get beyond this OpenTelemetry issue was autoregistration = false. When that broke stuff, I had to dig more and found the ExcludeTypeFromAutoRegistration, which allowed me to keep auto registration, but block off that one class in Otel that was causing the issue.
